pkg/scheduler/util/non_zero.go

```diff
@@ -34,45 +37,72 @@ const (
 	DefaultMilliCPURequest int64 = 100 // 0.1 core
 	// DefaultMemoryRequest defines default memory request size.
 	DefaultMemoryRequest int64 = 200 * 1024 * 1024 // 200 MB
+	// DefaultEphemeralStorage defines default empheral storage request size.
+	DefaultEphemeralStorage int64 = 0 // 0 Gi
 )
 
 // GetNonzeroRequests returns the default cpu and memory resource request if none is found or
 // what is provided on the request.
-func GetNonzeroRequests(requests *v1.ResourceList) (int64, int64) {
-	return GetNonzeroRequestForResource(v1.ResourceCPU, requests),
-		GetNonzeroRequestForResource(v1.ResourceMemory, requests)
+func GetNonzeroRequests(pod *v1.Pod) (int64, int64) {
+	return GetNonzeroRequestForResource(pod, v1.ResourceCPU),
+		GetNonzeroRequestForResource(pod, v1.ResourceMemory)
 }
 
 // GetNonzeroRequestForResource returns the default resource request if none is found or
 // what is provided on the request.
-func GetNonzeroRequestForResource(resource v1.ResourceName, requests *v1.ResourceList) int64 {
-	switch resource {
-	case v1.ResourceCPU:
-		// Override if un-set, but not if explicitly set to zero
-		if _, found := (*requests)[v1.ResourceCPU]; !found {
-			return DefaultMilliCPURequest
-		}
-		return requests.Cpu().MilliValue()
-	case v1.ResourceMemory:
-		// Override if un-set, but not if explicitly set to zero
-		if _, found := (*requests)[v1.ResourceMemory]; !found {
-			return DefaultMemoryRequest
+func GetNonzeroRequestForResource(pod *v1.Pod, resource v1.ResourceName) int64 {
+	podRequest := resourcehelper.GetResourceRequest(pod, resource)
+	resourceFound := true
+
+	// If Overhead is being utilized, add to the total requests for the pod
+	for i := range pod.Spec.Containers {
```

Comment:
Given that `resourcehelper.GetResourceRequest` has returned a summed amount for `resource` towards each container, I'd suggest just check if that returned amount is 0; if yes, return a default<Resource>Value directly. So we don't need to iterate each container here.

Reply:
Yep. My thought is also same with you before. But, there is a corner case
https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/types/node_info_test.go#L581

In the unit test above, it requests the CPU resource and overhead resources option, the memory request is nil. When we calculate the podResourceRequest through calling the `resourcehelper.GetResourceRequest`, it will ignore the overhead value of memory. In that case, the memory request was calculated as 0. So we need to consider both of the default value in memory request and overhead value of memory in the for-loop

Actually, I asked this question to the sig-api-machinery(https://kubernetes.slack.com/archives/C0EG7JC6T/p1585995903228600), from the implementation perspective of PodOverhead, we shouldn't add the podOverhead to the resource request, it should be added to the resource limit.

https://github.com/kubernetes/kubernetes/commit/80ee072b85fb8616b5209cfcafc291ccbf868657
```
Pod and burstable QoS cgroups should take overhead of running a sandbox
into account if the PodOverhead feature is enabled. These helper
functions are utilized by Kubelet for sizing the pod and burstable QoS
cgroups.

Pod overhead is added to resource requests, regardless of the initial
request values. A particular resource pod overhead is only added to a
resource limit if a non-zero limit already existed.
```

But, I think the kube-scheduler has own logic of handling resource request. So, we may need to continue to handle.

Reply:
Oh, the `!requestQuantity.IsZero()` logic is really surprising to me - IMO the overhead value should be added to total request **unconditionally**.

@egernst Any background on not adding overhead to total request if the request is 0?

Reply:
kube-scheduler doesn't care about limits, just requests. And according to the commit message, we should always add it to the request.

Reply:
Can you resolve similar obsolete discussion comments like this? (click the "resolve" button)